src/v/http/rest_client/rest_entity.h

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "http/request_builder.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/sstring.hh>
+
+#include <absl/container/flat_hash_map.h>
+#include <boost/beast/http/verb.hpp>
+
+namespace http::rest_client {
+
+// A rest entity represents an API endpoint and provides methods for CRUD
+// operations on it. For example given a REST resource /person, the rest entity
+// person{} is expected to provide create, retrieve, update and delete
+// operations relative to that REST resource. This does not however represent a
+// single person object such as /person/1.
+class rest_entity {
+public:
+    // The root url is used as the parent path for a rest entity. So given a
+    // REST resource /api/v1/person, the root url should be specified as
+    // /api/v1.
+    explicit rest_entity(std::string_view root_url);
+
+    // The name of the resource which is appended to the root url to generate
+    // the full path. In the canonical person example, it should be "person".
+    // Along with a root url of "/api/v1" this will result in calls to the path
+    // "/api/v1/person"
+    virtual ss::sstring resource_name() const = 0;
```

Comment:
I'm curious why this is a virtual method instead of accepting a string in the constructor for the resource name. This method takes no parameters, so the obvious thing is that that the return value is constant. Is the idea that derived classes might use some internal state to update the name between invocations?

Reply:
Yeah, the idea is to allow derived classes to be able to generate resource names using their own c-tor arguments, so ultimately a resource url can be generated for them. 

For example https://github.com/redpanda-data/redpanda/blob/7ac93ccdfa1f06e583fc39fd44e6213b98688b48/src/v/iceberg/rest_client/entities.cc#L26-L28 : a table entity is nested within a namespace, so the entity takes a namespace name (which is not constant) as c-tor arg and then generates a resource name for it like `namespace/{}/tables`. Then we can execute operations like list tables in namespace `x` by constructing a table entity with namespace `x` and issuing it's GET call.

A simpler implementation here would be some kind of public const virtual string field which can be overridden in subclasses but I couldn't see a way to implement it so I used this const getter approach. I'm open to suggestions if there is an easier to understand way to implement this.